contracts/FeralfileArtworkV4_2.sol

```diff
@@ -0,0 +1,339 @@
+        super._checkContractOwnedToken();
+        validateSaleData(saleData_);
+
+        saleData_.payByVaultContract
+            ? vault.payForSale(r_, s_, v_, saleData_)
+            : require(
+                saleData_.price == msg.value,
+                "FeralfileExhibitionV4: invalid payment amount"
+            );
+
+        bytes32 message = keccak256(
+            abi.encode(block.chainid, address(this), saleData_)
+        );
+
+        if (!isValidSignature(message, r_, s_, v_)) {
+            revert InvalidSignature();
+        }
+
+        uint256 itemRevenue;
+        if (saleData_.price > saleData_.cost) {
+            itemRevenue =
+                (saleData_.price - saleData_.cost) /
+                saleData_.tokenIds.length;
+        }
+
+        uint256 distributedRevenue;
+        uint256 platformRevenue;
+        uint256 coinsTransferAmount = 0;
+        for (uint256 i = 0; i < saleData_.tokenIds.length; i++) {
+            // send NFT
+            _safeTransfer(
+                address(this),
+                saleData_.destination,
+                saleData_.tokenIds[i],
+                ""
+            );
+            // distribute royalty
+            RevenueShare[] memory revenueShares = saleData_.revenueShares[i];
+            uint256 remainingRev = itemRevenue;
+
+            // deduct advances payment from revenue
+            for (
+                uint256 j = 0;
+                j < revenueShares.length && remainingRev > 0;
+                j++
+            ) {
+                uint256 remainingAdvanceAmount = advances[
+                    revenueShares[j].recipient
+                ];
+                uint256 rev = remainingAdvanceAmount >= remainingRev
+                    ? remainingRev
+                    : remainingAdvanceAmount;
+                platformRevenue += rev;
+                advances[revenueShares[j].recipient] -= rev;
+                remainingRev -= rev;
+            }
+
+            // distribute revenue
+            if (remainingRev > 0) {
+                for (uint256 j = 0; j < revenueShares.length; j++) {
+                    address recipient = revenueShares[j].recipient;
+                    uint256 rev = (remainingRev * revenueShares[j].bps) / 10000;
+                    if (recipient == costReceiver) {
+                        platformRevenue += rev;
+                        continue;
+                    }
+                    distributedRevenue += rev;
+                    payable(recipient).transfer(rev);
+                }
+            }
+
+            TokenInfo memory tokenInfo = _tokenInfos[saleData_.tokenIds[i]];
+            coinsTransferAmount += tokenInfo.coinAmount;
+
+            emit BuyArtwork(saleData_.destination, saleData_.tokenIds[i]);
+        }
+
+        // Call the transfer function of the ERC20 contract
+        if (coinsTransferAmount > 0 && _erc20Token.balanceOf(address(this)) > coinsTransferAmount * 10**18) {
```

Comment:
if possible, please create a private function to get exact amount of coin instead of putting multiple magic values. 
Reference [here](https://github.com/bitmark-inc/feralfile-exhibition-smart-contract/pull/41/files#r1616986156).